### PR TITLE
Mastermind

### PR DESCRIPTION
**Pull request by cayenwang**
Implemented Mastermind

### CODE REVIEW

File: src/main/java/games/GameType.java [hunk 1]
```diff
@@ -200,7 +206,15 @@ public enum GameType {
     ChineseCheckers(2, 6,
             Arrays.asList(Strategy, Abstract),
             Arrays.asList(GridMovement),
-            CCGameState.class, CCForwardModel.class, CCParameters.class, CCGUIManager.class);
+            CCGameState.class, CCForwardModel.class, CCParameters.class, CCGUIManager.class),
+    Mastermind(1,1,
+            Arrays.asList(Simple, Abstract, CodeBreaking),
+            Arrays.asList(),
+            MMGameState.class, MMForwardModel.class, MMParameters.class, MMGUIManager.class),
+    Cluedo(2,6,
```
Comment (hopshackle) [inline src/main/java/games/GameType.java:214]: The pull request should just contain Mastermind (and any core changes required to support it). Cluedo is still in development so should be kept in its own branch.

File: src/main/java/games/mastermind/MMGameState.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        return true;
+    }
+
+    @Override
+    protected double _getHeuristicScore(int playerId) {
+        return getGameScore(playerId);
+    }
+
+    @Override
+    public double getGameScore(int playerId) {
```
Comment (hopshackle) [inline src/main/java/games/mastermind/MMGameState.java:85]: getGameScore should be reserved for an actual game score as per the rules of the game - which you can argue in Mastermind is equal to the number of rows unused. (Or, that there is no concept of score, in which case always return 0).

The code here is for a (very reasonable) heuristic.

**Review by hopshackle — CHANGES_REQUESTED** [review]
Looks very good - main point is to remove the Cluedo code from this PR to avoid putting a half-finished game into the main repository.

**Comment by cayenwang** [comment]
I've implemented the changes. Not sure if I need to re-mark this PR as 'ready for review'

**Pull request merged**